cmd/osbuild-jobsite-builder/main.go

```diff
@@ -0,0 +1,443 @@
+	}
+}
+
+func (builder *Builder) RegisterHandler(h Handler) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if err := h(w, r); err != nil {
+			logrus.Fatal(err)
+		}
+	})
+}
+
+func (builder *Builder) HandleClaim(w http.ResponseWriter, r *http.Request) error {
+	builder.GuardState(StateClaim)
+
+	if r.Method != "POST" {
+		logrus.WithFields(
+			logrus.Fields{"method": r.Method},
+		).Fatal("Builder.HandleClaim: unexpected request method")
+	}
+
+	fmt.Fprintf(w, "%s", "done")
+
+	logrus.Info("Builder.HandleClaim: Done.")
+
+	builder.SetState(StateProvision)
+
+	return nil
+}
+
+func (builder *Builder) HandleProvision(w http.ResponseWriter, r *http.Request) (err error) {
+	builder.GuardState(StateProvision)
+
+	if r.Method != "PUT" {
+		return fmt.Errorf("Builder.HandleProvision: Unexpected request method.")
+	}
+
+	logrus.WithFields(logrus.Fields{"argBuildPath": argBuildPath}).Debug("Builder.HandleProvision: Opening manifest.json.")
+
+	dst, err := os.OpenFile(
+		path.Join(argBuildPath, "manifest.json"),
+		os.O_WRONLY|os.O_CREATE|os.O_EXCL,
+		0400,
+	)
+
+	defer func() {
+		if cerr := dst.Close(); cerr != nil {
+			err = cerr
+		}
+	}()
+
+	if err != nil {
+		return fmt.Errorf("Builder.HandleProvision: Failed to open manifest.json.")
+	}
+
+	logrus.Debug("Builder.HandleProvision: Writing manifest.json.")
+
+	_, err = io.Copy(dst, r.Body)
+
+	if err != nil {
+		return fmt.Errorf("Builder.HandleProvision: Failed to write manifest.json.")
+	}
+
+	w.WriteHeader(http.StatusCreated)
+
+	if _, err := w.Write([]byte(`done`)); err != nil {
+		return fmt.Errorf("Builder.HandleProvision: Failed to write response.")
+	}
+
+	logrus.Info("Builder.HandleProvision: Done.")
+
+	builder.SetState(StatePopulate)
+
+	return nil
+}
+
+func (builder *Builder) HandlePopulate(w http.ResponseWriter, r *http.Request) error {
+	builder.GuardState(StatePopulate)
+
+	if r.Method != "POST" {
+		return fmt.Errorf("Builder.HandlePopulate: unexpected request method")
+	}
+
+	w.WriteHeader(http.StatusOK)
+
+	if _, err := w.Write([]byte(`done`)); err != nil {
+		return fmt.Errorf("Builder.HandlePopulate: Failed to write response.")
+	}
+
+	logrus.Info("Builder.HandlePopulate: Done.")
+
+	builder.SetState(StateBuild)
+
+	return nil
+}
+
+func (builder *Builder) HandleBuild(w http.ResponseWriter, r *http.Request) error {
+	builder.GuardState(StateBuild)
+
+	if r.Method != "POST" {
+		return fmt.Errorf("Builder.HandleBuild: Unexpected request method.")
+	}
+
+	var buildRequest BuildRequest
+
+	var err error
+
+	if err = json.NewDecoder(r.Body).Decode(&buildRequest); err != nil {
+		return fmt.Errorf("HandleBuild: Failed to decode body.")
+	}
+
+	if Build != nil {
+		logrus.Fatal("HandleBuild: Build started but Build was non-nil.")
+	}
+
+	args := []string{
+		"--store", path.Join(argBuildPath, "store"),
+		"--cache-max-size", "0",
+		"--output-directory", path.Join(argBuildPath, "export"),
```

Comment:
Reading this again, it should really run osbuild in `--json` mode, it's needed to parse osbuild's output into an osbuild result.

Reply:
Done.